usr.bin/tee/tee.1
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)tee.1	8.1 (Berkeley) 6/6/93
.\" $FreeBSD: src/usr.bin/tee/tee.1,v 1.4.2.2 2001/08/16 13:17:06 ru Exp $
.\" $DragonFly: src/usr.bin/tee/tee.1,v 1.2 2003/06/17 04:29:32 dillon Exp $
.\"
.Dd June 6, 1993
.Dt TEE 1
.Os
.Sh NAME
.Nm tee
.Nd pipe fitting
.Sh SYNOPSIS
.Nm
.Op Fl ai
.Op Ar
.Sh DESCRIPTION
The
.Nm
utility copies standard input to standard output,
making a copy in zero or more files.
The output is unbuffered.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl a
Append the output to the files rather than
overwriting them.
.It Fl i
Ignore the
.Dv SIGINT
signal.
.El
.Pp
The following operands are available:
.Bl -tag -width file
.It Ar file
A pathname of an output
.Ar file .
.El
.Pp
The
.Nm
utility takes the default action for all signals,
except in the event of the
.Fl i
option.
.Sh EXIT STATUS
.Ex -std
.Sh STANDARDS
The
.Nm
utility is expected to be
.St -p1003.2
compatible.


usr.bin/tee/tee.c
/*
 * Copyright (c) 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * @(#) Copyright (c) 1988, 1993 The Regents of the University of California.  All rights reserved.
 * @(#)tee.c	8.1 (Berkeley) 6/6/93
 * $FreeBSD: src/usr.bin/tee/tee.c,v 1.4 1999/08/28 01:06:21 peter Exp $
 * $DragonFly: src/usr.bin/tee/tee.c,v 1.5 2005/02/18 16:33:53 liamfoy Exp $
 */

#include <sys/types.h>
#include <sys/queue.h>
#include <sys/stat.h>

#include <err.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

struct desc_list {
	SLIST_ENTRY(desc_list) link;
	int fd;
	const char *name;
};
SLIST_HEAD(, desc_list) desc_head;

static void	add(int, const char *);
static void	usage(void);

int
main(int argc, char **argv)
{
	struct desc_list *p;
	int n, fd, rval, wval, flags;
	char *bp, *buf;
	int append, ch, exitval;
#define	BSIZE (8 * 1024)

	append = 0;
	while ((ch = getopt(argc, argv, "ai")) != -1)
		switch(ch) {
		case 'a':
			append = 1;
			break;
		case 'i':
			if (signal(SIGINT, SIG_IGN) == SIG_ERR)
				err(1, "signal");
			break;
		default:
			usage();
		}
	argv += optind;
	argc -= optind;

	if ((buf = malloc(BSIZE)) == NULL)
		err(1, "malloc failed");

	add(STDOUT_FILENO, "stdout");

	if (append)
		flags = O_WRONLY|O_CREAT|O_APPEND;
	else
		flags = O_WRONLY|O_CREAT|O_TRUNC;

	for (exitval = 0; *argv != NULL; ++argv) {
		if ((fd = open(*argv, flags, DEFFILEMODE)) < 0) {
			warn("open failed: %s", *argv);
			exitval = 1;
		} else {
			add(fd, *argv);
		}
	}

	while ((rval = read(STDIN_FILENO, buf, BSIZE)) > 0) {
		SLIST_FOREACH(p, &desc_head, link) {
			n = rval;
			bp = buf;
			do {
				if ((wval = write(p->fd, bp, n)) == -1) {
					warn("write failed: %s", p->name);
					exitval = 1;
					break;
				}
				bp += wval;
			} while (n -= wval);
		}
	}
	if (rval < 0)
		err(1, "read failed");
	exit(exitval);
}

static void
usage(void)
{
	fprintf(stderr, "usage: tee [-ai] [file ...]\n");
	exit(1);
}

static void
add(int fd, const char *name)
{
	struct desc_list *p;

	p = malloc(sizeof(struct desc_list));
	if (p == NULL)
		err(1, "malloc failed");
	p->fd = fd;
	p->name = name;
	SLIST_INSERT_HEAD(&desc_head, p, link);
}
